packages/auto-tls/src/auto-tls.ts

```diff
@@ -0,0 +1,200 @@
+import { ClientAuth } from '@libp2p/http-fetch/auth'
+import { serviceDependencies, stop } from '@libp2p/interface'
+import { debounce } from '@libp2p/utils/debounce'
+import { isLoopback } from '@libp2p/utils/multiaddr/is-loopback'
+import { isPrivate } from '@libp2p/utils/multiaddr/is-private'
+import { QUICV1, TCP, WebSockets, WebSocketsSecure, WebTransport } from '@multiformats/multiaddr-matcher'
+import { Crypto } from '@peculiar/webcrypto'
+import * as x509 from '@peculiar/x509'
+import * as acmeClient from 'acme-client'
+import { base36 } from 'multiformats/bases/base36'
+import type { AutoTLSComponents, AutoTLSInit, AutoTLS as AutoTLSInterface } from './index.js'
+import type { PeerId, PrivateKey, Logger, TypedEventTarget, Libp2pEvents, AbortOptions, TLSCertificate } from '@libp2p/interface'
+import type { AddressManager } from '@libp2p/interface-internal'
+import type { DebouncedFunction } from '@libp2p/utils/debounce'
+import type { Multiaddr } from '@multiformats/multiaddr'
+
+const crypto = new Crypto()
+x509.cryptoProvider.set(crypto)
+
+type CertificateEvent = 'certificate:provision' | 'certificate:renew'
+
+export class AutoTLS implements AutoTLSInterface {
+  private readonly log: Logger
+  private readonly addressManager: AddressManager
+  private readonly privateKey: PrivateKey
+  private readonly peerId: PeerId
+  private readonly events: TypedEventTarget<Libp2pEvents>
+  private readonly forgeEndpoint: string
+  private readonly forgeDomain: string
+  private readonly acmeDirectory: string
+  private readonly clientAuth: ClientAuth
+  private readonly timeout: number
+  private readonly renewThreshold: number
+  private started: boolean
+  private shutdownController?: AbortController
+  public certificate?: TLSCertificate
+  private fetching: boolean
+  private readonly fetchCertificates: DebouncedFunction
+  private renewTimeout?: ReturnType<typeof setTimeout>
+
+  constructor (components: AutoTLSComponents, init: AutoTLSInit = {}) {
+    this.log = components.logger.forComponent('libp2p:certificate-manager')
+    this.addressManager = components.addressManager
+    this.privateKey = components.privateKey
+    this.peerId = components.peerId
+    this.events = components.events
+    this.forgeEndpoint = init.forgeEndpoint ?? 'registration.libp2p.direct'
```

Comment:
Endpoint implies a URL with path, no? 

Also suggestion to move these defaults into a constants file.

Reply:
The path is `/v1/_acme-challenge` - I omitted the path here to prevent people specifying `/v2/_acme-challenge` and expecting it to work.

We could make them add `https://` on the front if you think it's more intuitive, otherwise I guess we should add a note that the `/v1/..` bit will be appended?

Reply:
Yeah note also works.

Reply:
Done - note & constants file added